containers/unit-tests/start

```diff
@@ -22,6 +23,34 @@ if test -z "${docker:=$(which podman || which docker || true)}"; then
   exit 1
 fi
 
-set -ex
-exec ${docker} run --shm-size=512M --volume "${top_srcdir}":/source:ro \
-       ${ccarg:+"$ccarg"} $flags -- "$image" $cmd
+# restore image from Semaphore cache
+if [ -d "${SEMAPHORE_CACHE_DIR:-}" ]; then
+    image_esc="${image//\//_}"
+    image_esc="${image_esc//:/_}"
+    for f in $SEMAPHORE_CACHE_DIR/docker/$image_esc-*; do
+        [ -e "$f" ] || continue
+        echo "Restoring cached Docker image $f..."
+        ${docker} load < "$f"
+    done
+fi
+
+set -x
+RC=0
+${docker} run --shm-size=512M --volume "${top_srcdir}":/source:ro \
+       ${ccarg:+"$ccarg"} $flags -- "$image" $cmd || RC=$?
+
+# update Docker image in Semaphore cache
+if [ -d "${SEMAPHORE_CACHE_DIR:-}" ]; then
+    mkdir -p $SEMAPHORE_CACHE_DIR/docker
+    hash=$(docker images -q "$image")
+    cache_file="$SEMAPHORE_CACHE_DIR/docker/${image_esc}-$hash"
+    if ! [ -e "$cache_file" ]; then
+        # clean up older images
+        rm --verbose -f $SEMAPHORE_CACHE_DIR/docker/$image_esc-*
```

Comment:
First> How do you define "other"? We e. g. don't want the :i386 test to remove the cached :latest image and vice versa. We could hardcode the set of known/current images, but I don't like that much either TBH.

Second> The trick is that we need to pull an image to get its hash, so I can't do it in the loading section. I can of course do the cleanup outside of the if and filter out the current one, but that's not going to make much practical difference. Usually the old one becomes out of date *exactly when* you have to write a new one.

atime: semaphore saves/restores the cache in between VMs; at first sight it seems it does not get restored correctly:
```
Semaphore (57 min left): ~/cockpit $ stat .semaphore-cache/docker/cockpit_unit-tests_latest-7a48e3e45d2f 
  File: '.semaphore-cache/docker/cockpit_unit-tests_latest-7a48e3e45d2f'
  Size: 1413226496	Blocks: 2760216    IO Block: 4096   regular file
Device: fc00h/64512d	Inode: 138001      Links: 1
Access: (0600/-rw-------)  Uid: ( 1003/  runner)   Gid: ( 1003/  runner)
Access: 2019-10-01 17:05:03.284000000 +0000
Modify: 2019-10-01 15:57:15.000000000 +0000
Change: 2019-10-01 17:05:14.663035000 +0000
 Birth: -
Semaphore (57 min left): ~/cockpit $ date
Tue Oct  1 17:08:14 UTC 2019
```

I. e. it has ctime and atime when the VM started, instead of the earlier run when it was created and used.